lib/active_merchant/billing/gateways/cyber_source.rb

```diff
@@ -1193,28 +1213,105 @@ def add_partner_solution_id(xml)
         xml.tag!('partnerSolutionID', application_id)
       end
 
-      # Where we actually build the full SOAP request using builder
-      def build_request(body, options)
-        xsd_version = test? ? TEST_XSD_VERSION : PRODUCTION_XSD_VERSION
+      def sign_info(body)
+        canonicalized = canonicalization(body)
+        digest_value = OpenSSL::Digest::SHA256.digest(canonicalized)
 
         xml = Builder::XmlMarkup.new indent: 2
-        xml.instruct!
-        xml.tag! 's:Envelope', { 'xmlns:s' => 'http://schemas.xmlsoap.org/soap/envelope/' } do
-          xml.tag! 's:Header' do
-            xml.tag! 'wsse:Security', { 's:mustUnderstand' => '1', 'xmlns:wsse' => 'http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd' } do
+
+        xml.tag! 'ds:SignedInfo', { 'xmlns:ds' => NS_DS } do
+          xml.tag! 'ds:CanonicalizationMethod', { 'Algorithm' => C14N_ALGORITHM }
+          xml.tag! 'ds:SignatureMethod', { 'Algorithm' => RSA_SHA256_ALGORITHM }
+          xml.tag! 'ds:Reference', { 'URI' => '#Body' } do
+            xml.tag! 'ds:Transforms' do
+              xml.tag! 'ds:Transform', { 'Algorithm' => C14N_ALGORITHM }
+            end
+            xml.tag! 'ds:DigestMethod', { 'Algorithm' => SHA256_ALGORITHM }
+            xml.tag! 'ds:DigestValue', Base64.encode64(digest_value).strip
+          end
+        end
+
+        canonicalization(xml.target!)
+      end
+
+      def signature_element(xml, body)
+        signed_info = sign_info(body)
+        rsa_private_key = OpenSSL::PKey::RSA.new(@options[:private_key])
+        signature = rsa_private_key.sign(OpenSSL::Digest.new('SHA256'), signed_info)
+
+        xml.tag! 'ds:Signature', { 'xmlns:ds' => NS_DS } do
+          xml << signed_info
+          xml.tag! 'ds:SignatureValue', { 'xmlns:ds' => NS_DS } do
```

Comment:
Is it possible for someone to extract the private key from here?

Reply:
No, it is not possible to extract the private key from the signature. The private key is used to generate the signature, but the signature itself does not contain the private key. The signature is a cryptographic hash that can be verified using the corresponding public key, but it cannot be reversed to obtain the private key. However, it is possible to add the digest value and signature to the scrub method to prevent any information in the transcript from being exposed.

Reply:
If it won't expose any sensitive info I think it's fine. I will approve but can you just get double confirmation from Cybersource on this?

Reply:
Sure! While we wait for confirmation from Cybersource, I added the digest value and the signature to the scrub method.